cdap-security/src/main/java/co/cask/cdap/security/securestore/FileSecureStore.java

```diff
@@ -0,0 +1,451 @@
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.CertificateException;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import static java.nio.file.StandardCopyOption.ATOMIC_MOVE;
+
+/**
+ * File based implementation of secure store. Uses Java JCEKS based keystore.
+ *
+ * When the client calls a put, the key is put in the keystore. The system then flushes
+ * the keystore to the file system.
+ * During the flush, the current file is first written to temporary file (_NEW).
+ * If that is successful then the temporary file is renamed atomically to the secure store file.
+ * If anything fails during this process then the keystore reverts to the last successfully written file.
+ * The keystore is flushed to the filesystem after every put and delete.
+ */
+class FileSecureStore implements SecureStore, SecureStoreManager {
+  private static final Logger LOG = LoggerFactory.getLogger(FileSecureStore.class);
+
+  private static final String SCHEME_NAME = "jceks";
+  private static final Gson GSON = new Gson();
+
+  private final char[] password;
+  private final Path path;
+  private final Lock readLock;
+  private final Lock writeLock;
+  private final KeyStore keyStore;
+
+  FileSecureStore(CConfiguration cConf) throws IOException {
+    // Get the path to the keystore file
+    String pathString = cConf.get(Constants.Security.Store.FILE_PATH);
+    Path dir = Paths.get(pathString);
+    path = dir.resolve(cConf.get(Constants.Security.Store.FILE_NAME));
+
+    // Get the keystore password
+    password = cConf.get(Constants.Security.Store.FILE_PASSWORD).toCharArray();
+
+    keyStore = locateKeystore(path, password);
+    ReadWriteLock lock = new ReentrantReadWriteLock(true);
+    readLock = lock.readLock();
+    writeLock = lock.writeLock();
+  }
+
+  /**
+   * Stores an element in the secure store.
+   * @param name Name of the element to store
+   * @param data The data that needs to be securely stored
+   * @param properties Metadata associated with the data
+   * @throws IOException
+   */
+  @Override
+  public void put(String name, byte[] data, Map<String, String> properties) throws IOException {
+    SecureStoreMetadata meta = SecureStoreMetadata.of(name, properties);
+    SecureStoreData secureStoreData = new SecureStoreData(meta, data);
+    writeLock.lock();
+    try {
+      keyStore.setKeyEntry(name, new KeyStoreEntry(secureStoreData, meta), password, null);
+      // Attempt to persist the store.
+      flush();
+    } catch (KeyStoreException e) {
+      // We failed to store the key in the key store. Throw an IOException.
+      throw new IOException("Failed to store the key. ", e);
```

Comment:
when it is a non-IOException failure, would it be better to throw a more specific ``SecureStoreException``? Just checking if it would make sense for the client code to be able to easily distinguish between this and IOException and take appropriate action?